### ISSUE TITLE
Hyperparameter optimisation with Optuna (Hydra sweeper)

### ISSUE THREAD

**Issue by farhanferoz** [alan-turing-institute/autocast#377]
Current sweeps are Hydra multi-run grids (see SCRIPTS_AND_CONFIGS.md §3
and the `local_hydra/local_experiment/ablations/ae_compression/`
configs). Grid is fine for 1–2 axes; we should adopt
`hydra-optuna-sweeper` for higher-dimensional search with TPE / CMA-ES,
warm-starting, and (eventually) pruning.

### Proposal

- Add `hydra-optuna-sweeper` as a dev dependency.
- Add `hydra/sweeper: optuna` config group (one shared default + per-stage
  overrides for AE, EPD, processor-only).
- Wire the trial objective to a single scalar from `metrics.csv` (default
  `val_fcrps`; configurable per study).
- Document the workflow alongside the existing multi-run grid in
  `SCRIPTS_AND_CONFIGS.md`.

### Hyperparameters to consider (per stage)

Encoder/decoder (AE study):
- `encoder.latent_channels`, depth / widths, downsampling factor.
- Optimizer: `learning_rate`, `weight_decay`, warmup steps.

Processor (per family):
- Common: `learning_rate`, EMA decay, dropout, gradient clip, batch size.
- ViT/Swin: depth, heads, embed dim, patch size, attention dropout.
- FNO: modes, widths, depth.
- Flow matching: integration steps, sigma schedule.
- Diffusion: schedule, num steps, prediction parametrisation.
- Drifting / cohort: `n_samples`, `R_list`, kernel weights.

Data / training:
- `datamodule.batch_size`, `input_noise_injector.sigma`, rollout horizon
  during training.

### Minimum viable v1 — capped at 5 effective full-budget trainings

The smallest config that has the biggest impact within the cap. Pruning
and TPE are deliberately deferred; their value is marginal at this
budget and they add infrastructure cost.

**Two axes (no third).**
- `optimizer.learning_rate` — log-uniform `[1e-5, 1e-2]`.
- `optimizer.weight_decay` — log-uniform `[1e-6, 1e-2]`.

LR is the dominant lever empirically; WD couples with it; everything
else delivers smaller wins per trial at this budget.

**Sampler: `RandomSampler` (or `TPESampler` with `n_startup_trials=6`).**
At ≤ 12 trials TPE's KDE has too few observations to beat random in
published benchmarks. Start with `RandomSampler` for simplicity and
switch to TPE once the per-study trial count grows past ~25. The
official `hydra-optuna-sweeper` exposes sampler config directly, so the
upgrade is one YAML line.

**No pruner in v1.** The official `hydra-optuna-sweeper` does not
natively wire Optuna pruners; enabling them requires either manual
`trial.report` / `should_prune` inside the objective, or a community
fork (`DirkKuhn/hydra_optuna_pruning_sweeper`). Hyperband also costs
~4 × `n_startup_trials` before TPE engages, which alone exceeds the
5-cap. Defer to v2 when per-study budget grows.

**Per-trial budget.** Each Optuna trial trains at `trainer.max_time =
(1/3) × production max_time`. The 1/3-budget LR/WD *ranking* is mostly
stable; the *value* of the optimal LR can drift slightly (longer
horizons → smaller LR). Mitigation below.

**Budget allocation (5 effective full trainings = the cap):**
- **8 nominal HPO trials × 1/3 prod budget ≈ 2.7 effective.**
- **Top-2 candidates refit at full production budget = 2.0 effective.**
- **Total ≈ 4.7 effective.** Refitting the top *two* (not just the top
  one) catches the case where 1/3-budget rank ≠ full-budget rank.

**Objective:** single scalar, lower is better. `val_fcrps` for
probabilistic stacks; `val_rmse` for deterministic.

**Storage:** `sqlite:///optuna_${study_name}.db` on shared FS so the
study survives interrupts and supports later parallel workers.

**Stage ordering.** AE study → freeze winning AE → processor study on
top. Joint AE + processor HPO is combinatorially worse and not
affordable here.

**YAML stub (validated against current `hydra-optuna-sweeper` API):**

```yaml
# hydra/sweeper/optuna_min.yaml
defaults: [optuna]

sampler:
  _target_: optuna.samplers.RandomSampler   # swap → TPESampler at ~25+ trials/study
  seed: 42

n_trials: 8
n_jobs: 1                                   # raise on Baskerville
direction: minimize
study_name: ???
storage: sqlite:///optuna_${hydra.sweeper.study_name}.db

params:
  optimizer.learning_rate: tag(log, interval(1e-5, 1e-2))
  optimizer.weight_decay:  tag(log, interval(1e-6, 1e-2))

# Per-trial shortened budget set on the launching command, e.g.:
#   trainer.max_time=00:00:40:00  (≈ 1/3 of production at Stage-2 scale)
```

**Worked example (Stage-2-directional scale, ~2 h production):**
- HPO per-trial budget: `trainer.max_time=00:00:40:00`.
- 8 trials × 40 min ≈ 5.3 h.
- Top-2 refit at 2 h each = 4 h.
- Total study wall ≈ 9 h (≈ overnight on one GPU; ~3 h at `n_jobs=4`).

**v2 upgrade path** (once per-study budget grows past ~25 effective):
- Switch `RandomSampler` → `TPESampler(n_startup_trials=10)`.
- Add Hyperband pruning via the `hydra_optuna_pruning_sweeper` fork or
  manual `trial.report` wiring.
- Add a third axis (architecture width / modes; loss-specific knob).

### Open questions

1. **Upstream plugin vs community fork.** Official
   `hydra-optuna-sweeper` does not support pruners natively; the
   `DirkKuhn/hydra_optuna_pruning_sweeper` fork does. Decide upfront
   whether v2 pruning goes via the fork (cleaner config, dependency
   risk) or via manual `trial.report` / `should_prune` inside the
   objective (no extra dep, more code).
2. **Objective metric.** `val_fcrps` for probabilistic; what for
   deterministic? `val_rmse`? Composite?
3. **Search budget at scale.** v1 is capped at 5 effective; what's the
   right cap for a "real" study once we have headroom — 25? 50?
4. **Multi-stage HPO.** AE and processor sweeps are interdependent (AE
   choice shapes processor latent space). v1 fixes the ordering
   (AE → freeze → processor); confirm this matches team intent.
5. **Parallelism.** Optuna supports distributed studies via a shared
   storage backend (SQLite / RDB). Pick a storage convention that works
   on Baskerville (shared filesystem) and locally.
6. **Reproducibility / disk.** Each trial gets its own `run_id`; do we
   keep per-trial outputs or aggressively prune to save disk?
7. **Categorical vs continuous.** Architecture knobs (depth, modes) are
   discrete; mix with continuous LR/dropout via Optuna's mixed-space
   support.

Related: #366 (loss refactor — may affect what's tunable per loss),
#371 (perf — independent but trial throughput matters here).
